File: Python/17.py
#  17. Letter Combinations of a Phone Number

#  Given a string containing digits from 2-9 inclusive,
#  return all possible letter combinations that the number could represent.

#  A mapping of digit to letters (just like on the telephone buttons) is given below.
#  Note that 1 does not map to any letters.
from typing import List


class Solution:
    key_map = {
        "2": "abc",
        "3": "edf",
        "4": "ghi",
        "5": "jkl",
        "6": "mno",
        "7": "pqrs",
        "8": "tuv",
        "9": "wxyz",
    }

    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
        if len(digits) == 1:
            return list(Solution.key_map[digits[0]])
        results = []
        for result in self.letterCombinations(digits[1:]):
            for letter in Solution.key_map[digits[0]]:
                results.append(letter + result)
        return results


if __name__ == "__main__":
    from util import Test

    t = Test(Solution().letterCombinations)

    t.same_elements(["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"], "23")

File: cpluscplus/17.cpp
// 17. Letter Combinations of a Phone Number
// Given a string containing digits from 2-9 inclusive,
//  return all possible letter combinations that the number could represent.

// A mapping of digit to letters (just like on the telephone buttons) is
//  given below. Note that 1 does not map to any letters.

// Example:

// Input: "23"
// Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

class Solution
{
private:
    vector<string> number2string = ["abc", ] vector<string> help(vector<string> ans, string digits, int index) {
        vector<string> ans = {};
        if (index == (digits.size() - 1))
        {
        }
    }

    public : vector<string>
             letterCombinations(string digits)
    {
        vector<string> ans = {};
        return help(ans, digits, 0);
    }
};
int main()
{
    Solution s = Solution();
    string d;
    vector<string> ans;
    d = "23";
    ans = s.letterCombinations(d);
    cout << ans << endl;
}
